### PR TITLE
Persist form8.

### LINKED ISSUE #350: Record form 8 inputs in postgres for data analysis

### PR DESCRIPTION
**Pull request by amprokop**
Adds table for Form 8 data, persists it, refactors the Certifications controller and class, Form8 model, and Appeals.

Fixes #350 and #321. 

Test plan:
Lots of new unit tests
Lots of manual testing — creating a certification, then filling it in, advancing to the PDF preview, then closing the browser and returning to the certification page, which is still populated with the correct info.
Unit testing caught a few big bugs during development. I endeavored to add tests for each new significant chunk of code i wrote.

### CODE REVIEW

**Comment by amprokop** [comment]
@shanear — can you look at this schema for a sanity check before i continue working on this?

File: app/models/form8.rb [hunk 1]
```diff
@@ -48,61 +50,117 @@ class Form8
...
+    :_initial_other_notification_date,
+    :_initial_representative_name,
+    :_initial_representative_type,
+    :_initial_hearing_requested,
+    :_initial_contested_claims_procedures_applicable,
+    :_initial_contested_claims_requirements_followed,
+    :_initial_soc_date
+  ].freeze
+
+  def set_version
```
Comment (amprokop) [inline app/models/form8.rb]: see if i can put this back in initialize, had to take it out when we were passing in params due to whitelisting errors

File: app/models/form8.rb [hunk 2]
```diff
@@ -48,61 +50,117 @@ class Form8
...
+    self.version = SERIALIZATION_VERSION unless version
+  end
+
+  def save_pdf!
+    Form8.pdf_service.save_pdf_for!(self)
+  end
+
+  def update_from_appeal(appeal)
```
Comment (amprokop) [inline app/models/form8.rb]: rubocop will not like this

File: app/views/certifications/show.html.erb [hunk 3]
```diff
@@ -8,7 +8,7 @@
 </div>
 
 <iframe class="cf-doc-embed cf-app-segment" title="Form8 PDF"
-  src="<%= pdfjs.full_path(file: pdf_certification_path(id: form8.id, time: Time.now.to_i)) %>"></iframe>
+  src="<%= pdfjs.full_path(file: pdf_certification_path(id: form8.vacols_id, time: Time.now.to_i)) %>"></iframe>
```
Comment (amprokop) [inline app/views/certifications/show.html.erb]: form8 id is now the db id, not the vacols id. Q: we only want to permit one certification per vacols id, correct?

File: config/routes.rb [hunk 4]
```diff
@@ -16,6 +16,9 @@
     post 'confirm', on: :member
     get 'cancel', on: :member
   end
+
+  patch "certifications" => "certifications#create"
```
Comment (amprokop) [inline config/routes.rb]: it was necessary to add this since on the create action, we're now updating a resource, not creating it. rails will do a PATCH anyway since it knows that the form8 has been persisted, so we need to make the route handle that. not sure if there's a more semantic way to write this.

File: spec/models/certification_spec.rb [hunk 5]
```diff
@@ -129,27 +129,28 @@
...
-    context "when no cached form8 exists" do
-      before do
-        Rails.cache.write("TEST_form8", nil)
-      end
-
-      it "returns a new form8" do
-        expect(subject.file_number).to eq("VB12")
-      end
-    end
+    # TODO alex: certification.form8 no longer takes an argument. rewrite this block.
```
Comment (amprokop) [inline spec/models/certification_spec.rb]: this would be a good place to write tests showing that the certification is restored from the db i think

File: app/controllers/certifications_controller.rb [hunk 6]
```diff
@@ -14,22 +14,38 @@ def on_vbms_error
...
-    @form8 = @certification.form8(form8_cache_key)
+    # VACOLS data will be fetched fresh every time.
+    # However, visiting "/certifications/new/123C" will
+    # bring the user to the certification with VACOLS id "123C"
+    # if it already exists.
+    #
+    # Update the form8 that belongs to this certification
+    # with the newest data.
+    certification.form8.update_from_appeal(appeal)
+    @form8 = certification.form8
```
Comment (amprokop) [inline app/controllers/certifications_controller.rb]: need this line because `form_for` takes `@form8`

**Comment by amprokop** [comment]
TODO: add tests for `_initial_` fields, fix the date PDF display bug, verify that a back button test exists and create if not

**Comment by amprokop** [comment]
@shanear — this passes all the tests/rubocop/etc and is ready for another look!

File: spec/models/certification_spec.rb [hunk 7]
```diff
@@ -144,33 +144,76 @@
...
+        end
       end
     end
   end
 
+  context "#form8" do
+    # TODO: alex: certification.form8 no longer takes an argument. rewrite this block.
+    subject { certification.form8 }
+
+    # TODO: alex: rewrite these tests to use db instead of cache?
```
Comment (shanear) [inline spec/models/certification_spec.rb]: we're going to write these to test for the `find_or_create_by` logic?

File: spec/models/certification_spec.rb [hunk 8]
```diff
@@ -144,33 +144,76 @@
...
 
-    context "when no cached form8 exists" do
-      before do
-        Rails.cache.write("TEST_form8", nil)
-      end
+      context "when resuming a certification" do
+        it "does not update a form8's appeal data if form8 has been updated less than 48 hours ago" do
+          cert = Certification.new(vacols_id: "4949")
+          form = double
+          allow(cert).to receive(:form8).and_return(form)
```
Comment (shanear) [inline spec/models/certification_spec.rb]: lots of stubbing here. It's ok if we just get the database involved at this point.

File: spec/models/certification_spec.rb [hunk 9]
```diff
@@ -144,33 +144,76 @@
...
 
-    context "when a form8 exists in the cache for the passed key" do
-      before do
-        Rails.cache.write("TEST_form8", Form8.new(vacols_id: "4949", file_number: "SAVED88").attributes)
-      end
+          cert.start!
 
-      it "returns the cached form8" do
-        expect(subject.file_number).to eq("SAVED88")
+          expect(form).to have_received(:update_from_appeal).once
```
Comment (shanear) [inline spec/models/certification_spec.rb]: generally want to stay away from tests that expect method calls. These can be brittle and narrow. at that point it's ok to get the DB involved.

File: spec/feature/save_certification_spec.rb [hunk 10]
```diff
@@ -114,40 +117,7 @@
     click_on "Preview Completed Form 8"
 
     visit "certifications/new/6666C"
-    expect(find_field("Full Veteran Name").value).to eq("Crockett, Davy")
-  end
-
-  scenario "Does not repopulate saved form for past serialization versions" do
```
Comment (shanear) [inline spec/feature/save_certification_spec.rb]: yay for tests not being needed!

File: spec/feature/confirm_certification_spec.rb [hunk 11]
```diff
@@ -45,10 +48,10 @@
   end
 
   scenario "Successful confirmation with certification record" do
-    certification = Certification.create!(vacols_id: "5555C")
     visit "certifications/5555C"
     click_on "Upload and certify"
 
+    certification = Certification.find_or_create_by_vacols_id("5555C")
```
Comment (shanear) [inline spec/feature/confirm_certification_spec.rb]: why is this line necessary?

File: config/routes.rb [hunk 12]
```diff
@@ -10,14 +10,16 @@
...
-  
+
   resources :sessions, only: [:new, :create]
   resources :certifications, path_names: { new: "new/:vacols_id" } do
     get 'pdf', on: :member
     post 'confirm', on: :member
     get 'cancel', on: :member
   end
 
+  patch "certifications" => "certifications#create"
```
Comment (shanear) [inline config/routes.rb]: why did you add this route?

File: app/views/certifications/new.html.erb [hunk 13]
```diff
@@ -100,7 +100,7 @@
     <%= f.yes_no_field :hearing_transcript_on_file,
           question_number: "10B", part: "2", label: "Is the hearing transcript on file?", secondary: true %>
 
-    <%= f.text_field :hearing_requested_explaination,
+    <%= f.text_field :hearing_requested_explanation,
```
Comment (shanear) [inline app/views/certifications/new.html.erb]: yay spell check!

File: spec/models/certification_spec.rb (hunk 9 above)
Comment (amprokop) [inline spec/models/certification_spec.rb]: tbh i spent ~2 hours trying to get a version of this test that uses the db to work. lots of things outside here are stubbed out — `:load_vacols_data`, `:find_or_create_by_vacols_id`, the appeal record— and due to the let blocks it's not easy to test this—`certification` creates a new cert every time, etc. i can definitely go back to the well on this if you think it's high value, but just warning that it might take a lot of time

File: spec/models/certification_spec.rb (hunk 8 above)
Comment (amprokop) [inline spec/models/certification_spec.rb]: same comment as above — i'm willing to revisit this if you think it's high value, but this didn't play nicely with the existing data setup/teardown in my experience. might take significant time to rewrite.

File: config/routes.rb (hunk 12 above)
Comment (amprokop) [inline config/routes.rb]: `"certifications#create"` is no longer a POST request — under the hood, `form_for` changed it to a `PATCH`, since the form8 is actually saved before the create route is hit now. this change manifests itself as a hidden input on the html form, not anything defined by me. rails magic. i can show you if you like.

File: spec/models/form8_spec.rb [hunk 14]
```diff
@@ -24,6 +36,54 @@
...
+
+      params = {}
+      date_fields.each do |date_field|
+        params[date_field] = "02/01/2005"
+      end
+
+      form8.update_from_string_params(params)
+
+      date_fields.each do |date_field|
+        expect(form8[date_field].class).to eq(Date)
```
Comment (amprokop) [inline spec/models/form8_spec.rb]: sure sounds good

File: app/views/certifications/new.html.erb (hunk 13 above)
Comment (amprokop) [inline app/views/certifications/new.html.erb]: awww yeah

File: spec/feature/save_certification_spec.rb (hunk 10 above)
Comment (amprokop) [inline spec/feature/save_certification_spec.rb]: 👍

File: spec/models/certification_spec.rb (hunk 7 above)
Comment (amprokop) [inline spec/models/certification_spec.rb]: yep, done and pushed!

File: app/controllers/certifications_controller.rb [hunk 15]
```diff
@@ -14,22 +14,38 @@ def on_vbms_error
...
   end
 
   def create
-    @form8 = Form8.from_string_params(params[:form8])
-    Rails.cache.write(form8_cache_key, @form8.attributes)
-    form8.save!
-    redirect_to certification_path(id: form8.id)
+    # Can't use controller params in model mass assignments without whitelisting. See:
+    # http://edgeguides.rubyonrails.org/action_controller_overview.html#strong-parameters
+    params.require(:form8).permit!
```
Comment (amprokop) [inline app/controllers/certifications_controller.rb]: i marked the security scan Medium finding for this line as a false positive.

**Comment by amprokop** [comment]
@shanear — addressed all your comments so far! wanna take another look?
